packages/runtime/container-runtime/src/summaryBlobProtocol.ts

```diff
@@ -0,0 +1,164 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IsoBuffer } from "@fluidframework/common-utils";
+
+/**
+ * This class represents the object instatiation of the header which is optionaly writen at the
+ * beginning of the binary blob. It supports conversion to / from the binary array.
+ * The binary representation is as follows :
+ * - HEADER ID : 16 bytes of unique identifier : f4e72832a5bd47d7a2f35ed47ed94a3c
+ * - HEADER CONTENT LENGTH : 4 bytes of length of the rest of header content (after the length field)
+ * - HEADER CONTENT :  HEADER CONTENT LENGTH bytes
+ */
+export class BlobHeader {
+    public static readonly ENCODING = "utf-8";
+    public static readonly HEADER_PREFIX = new Uint8Array([0xf4, 0xe7, 0x28, 0x32, 0xa5, 0xbd, 0x47, 0xd7, 0xa2
```

Comment:
@vladsud, @anthony-murphy - I had a couple of quick thoughts on alternatives for identifying compressed blobs.

The first was to move the compression info from the blob to the reference (e.g., introduce a SummaryType.LZ4).  I can imagine this might not be easy to do if we store refs to blobs w/o metadata (e.g., driver cache), but I thought I'd mention it.

The other idea is that compression is a container-wide switch.  When compression is enabled, all blobs have a metadata header, and therefore we no longer need to worry about detecting the presence of a header.

This would prevent us for enabling compression on pre-exisitng containers, but would allow us to use a simple scheme like 00 = Uncompressed, 01 = LZ4, etc.

Thoughts?

Reply:
> This would prevent us for enabling compression on pre-exisitng containers, but would allow us to use a simple scheme like 00 = Uncompressed, 01 = LZ4, etc.

This is however a big penalty (impossibility to enable compression on pre-existing container), it is expected that containers will grow during it's lifecycle and this would impose unnecessary migration.

Reply:
We think the alternatives are worth exploring for the long-term benefit of having extensible metadata on our blobs.

As you mention above, the second option (prepend binary metadata) would need a common library for the client and service to parse the blob metadata.

I noticed that you called out option #1 (tag blob as compressed one level up in the summary tree) in your original PR summary under 'Questions and Arguments'.  We should do some investigation into this approach and see if it's practical, as that is the way we would have designed it if we had the foresight as it avoids copying the blob to prepend the header.

Reply:
> We think the alternatives are worth exploring for the long-term benefit of having extensible metadata on our blobs.
> 
> As you mention above, the second option (prepend binary metadata) would need a common library for the client and service to parse the blob metadata.
> 

Or as @vladsud noted, at least the well documented protocol. This all in case, the service wants to decompress the blobs. In any case, I do not consider sharing library such a big challenge, do you?

> I noticed that you called out option #1 (tag blob as compressed one level up in the summary tree) in your original PR summary under 'Questions and Arguments'. We should do some investigation into this approach and see if it's practical, as that is the way we would have designed it if we had the foresight as it avoids copying the blob to prepend the header.

I mentioned the option #1 before I found out that there is not a perfect fit between upload and download API. We upload ISummaryTree and download single blobs where the metadata would not be available. We would probably need to extend the communication APIs to deliver metadata together with the blob. It looks to be big and breaking change with possibly many consequences. I do not see this to be merged in reasonable time when we will need it. Anyway, please, give us your opinion.
The copying of the buffer is indeed quite a challenge, I wonder whether there is something like growing buffer or memory mapping when building one buffer from another one.

Reply:
Unfortunately, I'm still working on forming an opinion on our side.  Hopefully before the US holiday next week.

I did have a couple of ideas for avoiding the copy:

1. Modify lz4js to accept a destination buffer as an argument to compress.  We can then alloc a buffer with space reserved for the header.  (You can create a buffer that is a smaller subset of an existing buffer to elide the reserved space.)

2. Create a variant of LZ4 that compresses the source buffer "in-place" starting from the end so that the saved bytes end up at the beginning of the buffer.  (If we did that, we should call it 4ZL. :-)

(BTW - I don't think we should do either of these in your initial commit.)

Reply:
I should add, I think this is the last thing to resolve before we can take the commit.

I haven't yet reviewed the buffer manipulation code as it might change pending the above decision, but if there's any feedback it'll be quick/mechanical to address.

Reply:
A quick update: I'm following up with @agarwal-navin about the strategy he uses to deploy the GC features, as we think it may be a good general pattern for evolving the runtime.

As I understand it, the pattern is:
- Gradually roll out the next version of the client with the feature disabled.
- One there is sufficient saturation of the next version, enable the feature.
  - Existing documents are still in the old format until loaded / re-summarized
  - On the next summary, we record that the document is in the new format

This pattern will be more challenging once we have incremental summaries, which is one reason we'd like to get general metadata/headers into blobs soon.

Reply:
PS - We also agree with @milanro 's assessment that the asymmetry in the blob API would make it challenging to move the metadata into the blob references (i.e., the SummaryType.LZ4 idea above).

Reply:
@milanro - Stepping through the code, it looks like ISummaryTree is retrieved via a GET to '/git/summaries/latest' without going through `IDocumentStorageService.readBlob()`.  It seems like we should be able to note in the ISummaryTree if blobs are expected to have headers.  Hopefully I can poke at this a bit more tomorrow.

@scarlettjlee - This may be of interest to you: in order to compress summary blobs we need to know if the blobs are compressed.  The least intrusive way to do this is to prefix a short metadata header to blobs to tell `readBlobs()` if it should decompress.  In order to know if `readBlobs()` should interpret the first few bytes as metadata, we're thinking of adding something like a `version` field to ISummaryTree.